### ISSUE TITLE
Apps launched by xdg-open can misbehave when called from FHS envs or wrapped programs

### ISSUE THREAD

**Issue by LunNova** [NixOS/nixpkgs#160923]
### Describe the bug
When calling `xdg-open` from a FHS env or wrapped app, it tries to launch the associated program directly while either having unexpected/unrelated environment variables like `LD_PRELOAD` set or worse being in an FHS env which is entirely missing some dependencies.

Ideally, `xdg-open` when used in a FHS env or wrapped package would use `xdg-desktop-portal` to launch apps instead, similarly to how flatpaks use this to open other applications outside of the package.

There is partial upstream support for this but it only applies in flatpaks:  
https://github.com/freedesktop/xdg-utils/blob/1a58bc28f6844898532daf9ee1bf6da7764955a9/scripts/xdg-open.in#L250  
https://github.com/freedesktop/xdg-utils/blob/d11b33ec7f24cfb1546f6b459611d440013bdc72/scripts/xdg-utils-common.in#L361-L363  
It also doesn't handle local files correctly.

An example script which [also handles files is here](https://github.com/LunNova/nixos-configs/blob/b6688e198ed915be798073436eef7398ce89b410/packages/xdg-open-with-portal/default.nix).

### Steps To Reproduce
One example of this issue, although there are many related issues in different contexts.

1. Launch lutris under KDE
2. Click a link in it or an app launched by it
3. KDE thinks firefox's .desktop item isn't owned by root as it's owned by nobody in the FHS sandbox and refuses to start it

### Expected behavior
The `xdg-open` package in NixOS should use `xdg-desktop-portal` to launch apps/urls/etc when possible as if it is inside a flatpak to avoid running apps in unexpected FHS envs or with unexpected env from wrappers.

### Screenshots
![Opening a link in FFXIV under lutris](https://user-images.githubusercontent.com/782440/154811559-4b8c6602-486a-455b-8328-3ba16c8dc4ca.png)
![Error if you click yes](https://user-images.githubusercontent.com/782440/154811594-3b6eddb2-42c4-41d0-9c49-2101397c0a01.png)

### Notify maintainers
@edolstra - xdg-utils maintainer entry
@Atemu @illegalprime @wentasah - work on bwrap FHS envs 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.19-xanmod1-tt, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0pre20220127_558c4ee`
 - channels(root): `"nixos-21.11.333896.a640d8394f3"`
 - channels(lun): `""`
 - nixpkgs: `/etc/nix/path/nixpkgs`
```

**Comment by LunNova** [comment]
Related issues:
* #78961, #124901 - Firefox failing to open when Discord/signal-desktop launches it with xdg-open due to wrapper adding incompatible libraries, definitely this issue
* #73077 - Related for sublime3
* #137871 #120228 - various issues where xdg-open silently fails, likely this issue but not confirmed
* #152059 - I thought this was due to missing +x originally, closed in favor of this issue.
* #60260 - General issue for wrapper env vars interfering

**Comment by LunNova** [comment]
Related discourse threads:

* https://discourse.nixos.org/t/opening-links-in-todoist-electron-app/17774
* https://discourse.nixos.org/t/making-xdg-open-more-resilient/16777

**Comment by Atemu** [comment]
> The `xdg-open` package in NixOS should use `xdg-desktop-portal` to launch apps/urls/etc when possible as if it is inside a flatpak to avoid running apps in unexpected FHS envs or with unexpected env from wrappers.

Before we reinvent any wheels, how does Flatpak solve this problem?

**Comment by LunNova** [comment]
[xdg-utils detects `$XDG_RUNTIME_DIR/flatpak-info`](https://github.com/freedesktop/xdg-utils/blob/d11b33ec7f24cfb1546f6b459611d440013bdc72/scripts/xdg-utils-common.in#L361-L363) existing and [uses `xdg-desktop-portal`](https://github.com/freedesktop/xdg-utils/blob/1a58bc28f6844898532daf9ee1bf6da7764955a9/scripts/xdg-open.in#L250). This implementation is incomplete and will only handle URLs but hopefully it wouldn't be too had to get local file support added.

I'm not sure if making `xdg-desktop-portal` think it's in a flatpak in general is safe or if it'd be better to get a more nix-specific toggle added.

**Comment by Atemu** [comment]
> I'm not sure if making `xdg-open` think it's in a flatpak in general is safe

I mean, it basically is?  

Is there a non-Flatpak-specific way to handle portals? I imagine Snap and AppImage must be doing it somehow too. If they also use the Flatpak method, we should use it too. If they use a more generic way, we should use that.

**Comment by PAI5REECHO** [comment]
So is this why links from Discord open in Chrome instead of my default browser? :thinking:

**Comment by LunNova** [comment]
@PAI5REECHO 

> So is this why links from Discord open in Chrome instead of my default browser? thinking

It's possible although it's probably not that, it's more likely an issue with the xdg associations.

If you use home-manager, try something like [xdg-mime-apps.nix](https://github.com/LunNova/nixos-configs/blob/4f8cd98af25c7a1cd20d62af602c393d3ce66a65/users/lun/xdg-mime-apps.nix)

**Comment by PAI5REECHO** [comment]
@LunNova 
> it's more likely an issue with the xdg associations

It's not that, firefox is the only thing I have set in my `.config/mimeapps.list`

**Comment by LunNova** [comment]
I've implemented a workaround for this as a nixos module which is exposed at `nixosModules.xdg-open-workaround` in `github:LunNova/nixos-configs/dev`.

Implementation at https://github.com/LunNova/nixos-configs/blob/dev/modules/exported/xdg-open-workaround.nix and https://github.com/LunNova/nixos-configs/blob/dev/packages/xdg-open-with-portal/default.nix

Is this approach worth upstreaming?

**Comment by ShamrockLee** [comment]
Is #176962 related?

**Comment by SuperSandro2000** [comment]
@LunNova yes please

**Comment by schlichtanders** [comment]
> I've implemented a workaround for this as a nixos module which is exposed at `nixosModules.xdg-open-workaround` in `github:LunNova/nixos-configs/dev`.
> 
> Implementation at https://github.com/LunNova/nixos-configs/blob/dev/modules/exported/xdg-open-workaround.nix and https://github.com/LunNova/nixos-configs/blob/dev/packages/xdg-open-with-portal/default.nix
> 
> Is this approach worth upstreaming?

I would be glad if this problem can be solved generically, as it effects many applications (for me for example zettlr and marktext). I haven't tried your workaround, but it looks good when looking at the files.

@LunNova How can I install your nixos module?

**Comment by LunNova** [comment]
Add `github:LunNova/nixos-configs/main` as a flake input:

```
inputs = {
  ...
  lunnova-nixos-configs.url = "github:LunNova/nixos-configs/main";
  ...
}
```

Add it as an argument to outputs, then add `lunnova-nixos-configs.nixosModules.xdg-open-workaround` into the modules list you're passing to lib.nixosSystem.

Not sure how best to do it if you don't use flakes.

I'll try to get a PR to include this in nixpkgs shortly.

**Comment by deifactor** [comment]
@LunNova I haven't completely tried your approach, but one thing I noticed is that on my system `ulimit -Sn` is `1024`, so just running `exec 9999< /some/file` fails. Maybe pick your favorite three-digit number as the fd?

**Comment by LunNova** [comment]
Been somewhat distracted, hope to make a PR for this tonight. (Yes, this comment is an accountability post for myself so I'll actually do it :))

**Comment by LunNova** [comment]
Draft PR up here: #187293

**Comment by LunNova** [comment]
@deifactor I swapped it to use bash's "new" (from 2009 lol) syntax for auto selecting an available FD instead, hopefully that works on your machine!

**Comment by deifactor** [comment]
Looks like it works (that is, `exec {foo}< /etc/passwd; echo $foo` echoes 10, so I assume this means the syntax is compatible). Thanks!

**Comment by deifactor** [comment]
Not commenting on the quality of the fix itself, more "how did we get here".

So if I understand correctly, the problem is that your FHS application can have an arbitrary environment, which get passed to xdg-open, which get in turn passed down to Firefox. On most distributions, this isn't a problem because you don't do weird stuff with environment variables (or if you do, you do so for your entire user session so you'll break firefox immediately), but nixpkgs' isolation means that the FHS application might have LD_PRELOAD, LD_LIBRARY_PATH, or some other weird thing set that isn't set when invoking it normally.

1. Doesn't any solution that involves gdbus (or any 'wrapper' program) run the risk of having the same issue if you set some env variable in a way that makes gdbus sad?
2. Can we fix this by just having xdg-open  clear out variables that are suspected to be problematic (`LD_*`)? Obviously having *any* fix is better than nothing, but it feels a little weird that this solution relies on the fact that portals happen to exist, if that makes sense.

**Comment by LunNova** [comment]
> Doesn't any solution that involves gdbus (or any 'wrapper' program) run the risk of having the same issue if you set some env variable in a way that makes gdbus sad?

Yes, but it has a limited scope as once you manage to make a dbus call you're talking to the portal which isn't inheriting your environment, and it's only one combination to test (your program's env/fhs + gdbus) instead of your program's env/fhs + any other program in your system it might launch.

> Can we fix this by just having xdg-open clear out variables that are suspected to be problematic (LD_*)? Obviously having any fix is better than nothing, but it feels a little weird that this solution relies on the fact that portals happen to exist, if that makes sense.

This works for wrapped programs but not FHS envs. FHS envs use namespaces/cgroups, you can't undo this just by clearing an environment variable.

If your solution only needs to work for wrapped programs, clearing a set of env variables works.

**Comment by deifactor** [comment]
Ah, I wasn't aware of that re: FHS envs. (How that all works is not exactly within my wheelhouse.)

**Comment by LunNova** [comment]
Opened #197118 with an alternative approach. (#187293 was the original which relied on xdg-open in PATH working or would cause mass rebuilds toggling the setting)

Looking for help getting either approach in.

**Comment by LunNova** [comment]
#197118 has been merged so `xdg.portal = { enable = true; xdgOpenUsePortal = true; }` should work around this issue.

Please report here if it causes any new bugs or does not resolve your issue, along with your desktop environment and the app involved.

**Comment by huantianad** [comment]
@LunNova Is there a way to have `xdg-open` still support `file://` URIs when using `xdgOpenUsePortal = true`?

**Comment by LunNova** [comment]
We can probably patch support in in nixpkgs' version. I still haven't heard anything from upstream about the last PR so it doesn't seem like which is inconvenient for getting it fixed for everyone :/

**Comment by huantianad** [comment]
Sounds good, is it a simple patch to `open_flatpak` that checks for the `file` protocol and uses `OpenURI.OpenFile`/`OpenURI.OpenDirectory` instead?

**Comment by huantianad** [comment]
As a temporary workaround for my system, I'm using [`flatpak-xdg-utils`](https://github.com/flatpak/flatpak-xdg-utils), which provides its own `xdg-open` that works with `file://`. 

<https://github.com/huantianad/nixos-config/commit/147cb7e91d75d4764eb031875b8e527e1927be21>

**Comment by VAWVAW** [comment]
I am experiencing #189851 while trying this with a standalone wm. Just in case someone has the same issue.

**Comment by MathiasSven** [comment]
It seems that when I tried using this option `xdgOpenUsePortal = true` it fixed my issue where I couldn't open links from within `vscode-fhs` however now I had problems using `xdg-open` command from the terminal normally, which also made so some python application wouldn't open an image because it relied on this command. As an example:
```
~ ✦ ❯ xdg-open background.png
(objectpath '/org/freedesktop/portal/desktop/request/1_11449/t',)

~ ✦ ❯ 
```
No image would open and I would just get that printed to `stderr`. Turning this option off and logging off and back on solved the issue but then I am back to not being able to open links in FHS. I am not really sure how but it seems this option messes up with my non FHS environment.

**Comment by lilyinstarlight** [comment]
I've opened PR #217590 with a fix for using `xdg-open` on filenames when `xdgOpenUsePortal = true`

**Comment by chenlijun99** [comment]
@lilyinstarlight Thanks for your patch. A few days ago it landed on NixOS unstable so I tried it.. It seems that it fixed opening folders with `xdg-open` but didn't fix opening files. I'm still having the same behaviour that @MathiasSven showed in https://github.com/NixOS/nixpkgs/issues/160923#issuecomment-1411872545

**Comment by chenlijun99** [comment]
TL;DR;

The `plasma-xdg-desktop-portal-kde` needs an additional `plasma-workspace` package as build input.

---

Ok. Got the reason.

It turns out that with the new patched `xdg-open` it is possible to open some files, while for some other files it does nothing.
I am using `plasma-xdg-desktop-portal-kde` as backend for `xdg-desktop-portal` and for every `xdg-open` invocation that "does nothing", I could see the following log from `plasma-xdg-desktop-portal-kde`. 

```
xdg-desktop-portal-kde[3318]: QQmlApplicationEngine failed to load component
xdg-desktop-portal-kde[3318]: qrc:/AppChooserDialog.qml:15:1: module "org.kde.plasma.w>
xdg-desktop-portal-kde[3318]: QObject::connect: Cannot connect (nullptr)::accept() to >
xdg-desktop-portal-kde[3318]: QObject::connect: Cannot connect (nullptr)::reject() to >
xdg-desktop-portal-kde[3318]: Failed to load dialog, cannot exec
```

I suspected that the files that worked were the files for which `plasma-xdg-desktop-portal-kde` was able to pick a default program to use to open the file, while for the files that didn't work `plasma-xdg-desktop-portal-kde` tried to open an application picker dialog but failed and printed those logs. Thus I tried to add the dependency (`plasma-workspace`) that is presumably missing to the `xdg-desktop-portal-kde` package. I did it via overlay.

```nix
let
  pkgs-unstable = import inputs.nixpkgs-unstable {
    system = pkgs.system;
  };
in {
  nixpkgs.overlays = [
    (final: prev: {
      xdg-desktop-portal-kde = pkgs-unstable.xdg-desktop-portal-kde.overrideAttrs (oldAttrs: rec {
        buildInputs = oldAttrs.buildInputs ++ [pkgs-unstable.plasma-workspace];
      });
    })
  ];
}
```

After applying this change, indeed `xdg-open` started to open an application picker for the files that previously didn't work.

**Comment by huantianad** [comment]
@chenlijun99 interesting, I haven't had any issues with opening files, but actually when opening directories, if dolphin is already open, but I use `xdg-open` to open a new folder that isn't already open in the existing dolphin, it'll instead open a completely new instance of dolphin instead of opening it in my existing dolphin window. I'd assume this is a dolphin bug, but I can't seem to find it online anywhere. Wanted to mention this in case anyone else had this same issue and had a fix.

**Comment by MathiasSven** [comment]
@chenlijun99 I assume this could just be PRed then no?

**Comment by Atemu** [comment]
Could those affected test the PR above?

**Comment by Eisfunke** [comment]
I made some notes during my efforts with this problem in the hope that it might be helpful:

- I experienced this with `element-desktop` from unstable, I couldn't open links.
- I tried various workarounds from this issue and other places, none worked.
- setting the option `xdgOpenUsePortal` didn't help.
  - like others, that broke `xdg-open` from the terminal for me as well.
  - I only get `(objectpath '/org/freedesktop/portal/desktop/request/1_168/t',)` in the terminal with nothing else happening when trying to run `xdg-open` on any file *or link*.
  - note: the number `168` in the output of `xdg-open` is only an example, the actual number varies. To be specific, on each invocation it gets incremented by 2. This includes clicking links in Element, so if I'd get 168 in xdg-open, click a link in Element, use xdg-open again, I'd get 172. Quite hilarious, honestly.
  - This is a similiar thing to what @MathiasSven described, though for me this happened on links *and* files and I links in Element still didn't work.
  - I attached some logs from `dbus-monitor` when running `xdg-open`.
  - using `xdg-utils` from unstable, which should include the patch from above, doesn't help either.
  - it changes one thing though: when I use `xdg-open` on a file which doesn't have any handlers available. With `xdg-utils` from 22.11 I only get the `objectpath` thingy, with unstable `xdg-utils` I get a "No Apps Available" window.
  - happens with `xdg-desktop-portal-gtk` from both unstable and 22.11.
- Only at that point I found the issue regarding element in specific (#120228), switched to element-desktop from 22.11 again, that fixed it.
    - It still doesn't work with `xdgOpenUsePortal` set.

I'm using wayland with sway and `xdg-desktop-portal-wlr` and `xdg-desktop-portal-gtk`. For reference, my [config](https://git.eisfunke.com/config/nixos).

Finally, some logs `dbus-monitor` spews out when running `xdg-open`. Interesting stuff seems to be mostly in the end. My observations: it seems to have found vivaldi-stable.desktop, which is indeed the desktop file it should use to launch the link.  Also, the final message is:

```
error_name=org.freedesktop.DBus.Error.ServiceUnknown reply_serial=777
   string "The name :1.346 was not provided by any .service files"
```

and 346 is the number that `xdg-open` showed in the `objectpath` message from above. So yeah, no idea what's happening there.

<details>
<summary>dbus logs</summary>

```
method call time=1681858691.949944 sender=:1.345 -> destination=org.freedesktop.DBus serial=1 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=Hello
method return time=1681858691.949962 sender=org.freedesktop.DBus -> destination=:1.345 serial=1 reply_serial=1
   string ":1.345"
signal time=1681858691.949973 sender=org.freedesktop.DBus -> destination=(null destination) serial=740 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string ":1.345"
   string ""
   string ":1.345"
signal time=1681858691.949990 sender=org.freedesktop.DBus -> destination=:1.345 serial=2 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
   string ":1.345"
method call time=1681858691.950061 sender=:1.345 -> destination=org.freedesktop.DBus serial=2 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=GetNameOwner
   string "org.gnome.SessionManager"
error time=1681858691.950070 sender=org.freedesktop.DBus -> destination=:1.345 error_name=org.freedesktop.DBus.Error.NameHasNoOwner reply_serial=2
   string "Could not get owner of name 'org.gnome.SessionManager': no such name"
signal time=1681858691.950322 sender=org.freedesktop.DBus -> destination=:1.345 serial=5 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameLost
   string ":1.345"
signal time=1681858691.950339 sender=org.freedesktop.DBus -> destination=(null destination) serial=741 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string ":1.345"
   string ":1.345"
   string ""
method call time=1681858691.970429 sender=:1.346 -> destination=org.freedesktop.DBus serial=1 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=Hello
method return time=1681858691.970447 sender=org.freedesktop.DBus -> destination=:1.346 serial=1 reply_serial=1
   string ":1.346"
signal time=1681858691.970458 sender=org.freedesktop.DBus -> destination=(null destination) serial=742 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string ":1.346"
   string ""
   string ":1.346"
signal time=1681858691.970472 sender=org.freedesktop.DBus -> destination=:1.346 serial=2 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
   string ":1.346"
method call time=1681858691.970575 sender=:1.346 -> destination=org.freedesktop.portal.Desktop serial=2 path=/org/freedesktop/portal/desktop; interface=org.freedesktop.DBus.Introspectable; member=Introspect
method return time=1681858691.970999 sender=:1.6 -> destination=:1.346 serial=772 reply_serial=2
   string "<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
                      "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
<!-- GDBus 2.74.3 -->
<node>
  <interface name="org.freedesktop.DBus.Properties">
    <method name="Get">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="s" name="property_name" direction="in"/>
      <arg type="v" name="value" direction="out"/>
    </method>
    <method name="GetAll">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="a{sv}" name="properties" direction="out"/>
    </method>
    <method name="Set">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="s" name="property_name" direction="in"/>
      <arg type="v" name="value" direction="in"/>
    </method>
    <signal name="PropertiesChanged">
      <arg type="s" name="interface_name"/>
      <arg type="a{sv}" name="changed_properties"/>
      <arg type="as" name="invalidated_properties"/>
    </signal>
  </interface>
  <interface name="org.freedesktop.DBus.Introspectable">
    <method name="Introspect">
      <arg type="s" name="xml_data" direction="out"/>
    </method>
  </interface>
  <interface name="org.freedesktop.DBus.Peer">
    <method name="Ping"/>
    <method name="GetMachineId">
      <arg type="s" name="machine_uuid" direction="out"/>
    </method>
  </interface>
  <interface name="org.freedesktop.portal.Inhibit">
    <method name="Inhibit">
      <arg type="s" name="window" direction="in"/>
      <arg type="u" name="flags" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="CreateMonitor">
      <arg type="s" name="window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="QueryEndResponse">
      <arg type="o" name="session_handle" direction="in"/>
    </method>
    <signal name="StateChanged">
      <arg type="o" name="session_handle"/>
      <arg type="a{sv}" name="state"/>
    </signal>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Background">
    <method name="RequestBackground">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Location">
    <method name="CreateSession">
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="Start">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <signal name="LocationUpdated">
      <arg type="o" name="session_handle"/>
      <arg type="a{sv}" name="location"/>
    </signal>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Notification">
    <method name="AddNotification">
      <arg type="s" name="id" direction="in"/>
      <arg type="a{sv}" name="notification" direction="in"/>
    </method>
    <method name="RemoveNotification">
      <arg type="s" name="id" direction="in"/>
    </method>
    <signal name="ActionInvoked">
      <arg type="s" name="id"/>
      <arg type="s" name="action"/>
      <arg type="av" name="parameter"/>
    </signal>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Screenshot">
    <method name="Screenshot">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="PickColor">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Account">
    <method name="GetUserInformation">
      <arg type="s" name="window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.NetworkMonitor">
    <method name="GetAvailable">
      <arg type="b" name="available" direction="out"/>
    </method>
    <method name="GetMetered">
      <arg type="b" name="metered" direction="out"/>
    </method>
    <method name="GetConnectivity">
      <arg type="u" name="connectivity" direction="out"/>
    </method>
    <method name="GetStatus">
      <arg type="a{sv}" name="status" direction="out"/>
    </method>
    <method name="CanReach">
      <arg type="s" name="hostname" direction="in"/>
      <arg type="u" name="port" direction="in"/>
      <arg type="b" name="reachable" direction="out"/>
    </method>
    <signal name="changed"/>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Print">
    <method name="Print">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="title" direction="in"/>
      <arg type="h" name="fd" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="PreparePrint">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="title" direction="in"/>
      <arg type="a{sv}" name="settings" direction="in"/>
      <arg type="a{sv}" name="page_setup" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Settings">
    <method name="ReadAll">
      <arg type="as" name="namespaces" direction="in"/>
      <arg type="a{sa{sv}}" name="value" direction="out"/>
    </method>
    <method name="Read">
      <arg type="s" name="namespace" direction="in"/>
      <arg type="s" name="key" direction="in"/>
      <arg type="v" name="value" direction="out"/>
    </method>
    <signal name="SettingChanged">
      <arg type="s" name="namespace"/>
      <arg type="s" name="key"/>
      <arg type="v" name="value"/>
    </signal>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.GameMode">
    <method name="QueryStatus">
      <arg type="i" name="pid" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="RegisterGame">
      <arg type="i" name="pid" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="UnregisterGame">
      <arg type="i" name="pid" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="QueryStatusByPid">
      <arg type="i" name="target" direction="in"/>
      <arg type="i" name="requester" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="RegisterGameByPid">
      <arg type="i" name="target" direction="in"/>
      <arg type="i" name="requester" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="UnregisterGameByPid">
      <arg type="i" name="target" direction="in"/>
      <arg type="i" name="requester" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="QueryStatusByPIDFd">
      <arg type="h" name="target" direction="in"/>
      <arg type="h" name="requester" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="RegisterGameByPIDFd">
      <arg type="h" name="target" direction="in"/>
      <arg type="h" name="requester" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <method name="UnregisterGameByPIDFd">
      <arg type="h" name="target" direction="in"/>
      <arg type="h" name="requester" direction="in"/>
      <arg type="i" name="result" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.RemoteDesktop">
    <method name="CreateSession">
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="SelectDevices">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="Start">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="NotifyPointerMotion">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="d" name="dx" direction="in"/>
      <arg type="d" name="dy" direction="in"/>
    </method>
    <method name="NotifyPointerMotionAbsolute">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="u" name="stream" direction="in"/>
      <arg type="d" name="x" direction="in"/>
      <arg type="d" name="y" direction="in"/>
    </method>
    <method name="NotifyPointerButton">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="i" name="button" direction="in"/>
      <arg type="u" name="state" direction="in"/>
    </method>
    <method name="NotifyPointerAxis">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="d" name="dx" direction="in"/>
      <arg type="d" name="dy" direction="in"/>
    </method>
    <method name="NotifyPointerAxisDiscrete">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="u" name="axis" direction="in"/>
      <arg type="i" name="steps" direction="in"/>
    </method>
    <method name="NotifyKeyboardKeycode">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="i" name="keycode" direction="in"/>
      <arg type="u" name="state" direction="in"/>
    </method>
    <method name="NotifyKeyboardKeysym">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="i" name="keysym" direction="in"/>
      <arg type="u" name="state" direction="in"/>
    </method>
    <method name="NotifyTouchDown">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="u" name="stream" direction="in"/>
      <arg type="u" name="slot" direction="in"/>
      <arg type="d" name="x" direction="in"/>
      <arg type="d" name="y" direction="in"/>
    </method>
    <method name="NotifyTouchMotion">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="u" name="stream" direction="in"/>
      <arg type="u" name="slot" direction="in"/>
      <arg type="d" name="x" direction="in"/>
      <arg type="d" name="y" direction="in"/>
    </method>
    <method name="NotifyTouchUp">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="u" name="slot" direction="in"/>
    </method>
    <property type="u" name="AvailableDeviceTypes" access="read"/>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.MemoryMonitor">
    <signal name="LowMemoryWarning">
      <arg type="y" name="level"/>
    </signal>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.OpenURI">
    <method name="OpenURI">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="uri" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="OpenFile">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="h" name="fd" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="OpenDirectory">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="h" name="fd" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Realtime">
    <method name="MakeThreadRealtimeWithPID">
      <arg type="t" name="process" direction="in"/>
      <arg type="t" name="thread" direction="in"/>
      <arg type="u" name="priority" direction="in"/>
    </method>
    <method name="MakeThreadHighPriorityWithPID">
      <arg type="t" name="process" direction="in"/>
      <arg type="t" name="thread" direction="in"/>
      <arg type="i" name="priority" direction="in"/>
    </method>
    <property type="i" name="MaxRealtimePriority" access="read"/>
    <property type="i" name="MinNiceLevel" access="read"/>
    <property type="x" name="RTTimeUSecMax" access="read"/>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.DynamicLauncher">
    <method name="Install">
      <arg type="s" name="token" direction="in"/>
      <arg type="s" name="desktop_file_id" direction="in"/>
      <arg type="s" name="desktop_entry" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
    </method>
    <method name="PrepareInstall">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="name" direction="in"/>
      <arg type="v" name="icon_v" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="RequestInstallToken">
      <arg type="s" name="name" direction="in"/>
      <arg type="v" name="icon_v" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="s" name="token" direction="out"/>
    </method>
    <method name="Uninstall">
      <arg type="s" name="desktop_file_id" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
    </method>
    <method name="GetDesktopEntry">
      <arg type="s" name="desktop_file_id" direction="in"/>
      <arg type="s" name="contents" direction="out"/>
    </method>
    <method name="GetIcon">
      <arg type="s" name="desktop_file_id" direction="in"/>
      <arg type="v" name="icon_v" direction="out"/>
      <arg type="s" name="icon_format" direction="out"/>
      <arg type="u" name="icon_size" direction="out"/>
    </method>
    <method name="Launch">
      <arg type="s" name="desktop_file_id" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
    </method>
    <property type="u" name="SupportedLauncherTypes" access="read"/>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Wallpaper">
    <method name="SetWallpaperURI">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="uri" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="SetWallpaperFile">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="h" name="fd" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Camera">
    <method name="AccessCamera">
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="OpenPipeWireRemote">
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="h" name="fd" direction="out"/>
    </method>
    <property type="b" name="IsCameraPresent" access="read"/>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Device">
    <method name="AccessDevice">
      <arg type="u" name="pid" direction="in"/>
      <arg type="as" name="devices" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.PowerProfileMonitor">
    <property type="b" name="power-saver-enabled" access="read"/>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.ScreenCast">
    <method name="CreateSession">
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="SelectSources">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="Start">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="OpenPipeWireRemote">
      <arg type="o" name="session_handle" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="h" name="fd" direction="out"/>
    </method>
    <property type="u" name="AvailableSourceTypes" access="read"/>
    <property type="u" name="AvailableCursorModes" access="read"/>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Email">
    <method name="ComposeEmail">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.Trash">
    <method name="TrashFile">
      <arg type="h" name="fd" direction="in"/>
      <arg type="u" name="result" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.ProxyResolver">
    <method name="Lookup">
      <arg type="s" name="uri" direction="in"/>
      <arg type="as" name="proxies" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
  <interface name="org.freedesktop.portal.FileChooser">
    <method name="OpenFile">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="title" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="SaveFile">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="title" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <method name="SaveFiles">
      <arg type="s" name="parent_window" direction="in"/>
      <arg type="s" name="title" direction="in"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <property type="u" name="version" access="read"/>
  </interface>
</node>
"
method call time=1681858691.971556 sender=:1.346 -> destination=org.freedesktop.portal.Desktop serial=3 path=/org/freedesktop/portal/desktop; interface=org.freedesktop.portal.OpenURI; member=OpenURI
   string ""
   string "https://nixos.org"
   array [
   ]
method call time=1681858691.971733 sender=:1.6 -> destination=org.freedesktop.DBus serial=773 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=GetConnectionUnixProcessID
   string ":1.346"
method return time=1681858691.971746 sender=org.freedesktop.DBus -> destination=:1.6 serial=276 reply_serial=773
   uint32 55581
method return time=1681858691.971934 sender=:1.6 -> destination=:1.346 serial=774 reply_serial=3
   object path "/org/freedesktop/portal/desktop/request/1_346/t"
signal time=1681858691.972200 sender=org.freedesktop.DBus -> destination=:1.346 serial=6 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameLost
   string ":1.346"
signal time=1681858691.972211 sender=org.freedesktop.DBus -> destination=(null destination) serial=743 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string ":1.346"
   string ":1.346"
   string ""
method call time=1681858691.973147 sender=:1.6 -> destination=:1.8 serial=775 path=/org/freedesktop/impl/portal/PermissionStore; interface=org.freedesktop.impl.portal.PermissionStore; member=Lookup
   string "desktop-used-apps"
   string "x-scheme-handler/https"
error time=1681858691.973240 sender=:1.8 -> destination=:1.6 error_name=org.freedesktop.portal.Error.NotFound reply_serial=775
   string "No entry for x-scheme-handler/https"
signal time=1681858691.975112 sender=:1.6 -> destination=(null destination) serial=776 path=/org/gtk/gio/DesktopAppInfo; interface=org.gtk.gio.DesktopAppInfo; member=Launched
   array of bytes "/etc/profiles/per-user/eisfunke/share/applications/vivaldi-stable.desktop" + \0
   string ""
   int64 55590
   array [
      string "https://nixos.org"
   ]
   array [
      dict entry(
         string "origin-prgname"
         variant             array of bytes "/nix/store/90x9iwapqqm28w6nli53256i18l5jb04-xdg-desktop-portal-1.15.0/libexec/xdg-desktop-portal" + \0
      )
      dict entry(
         string "origin-pid"
         variant             int64 2084
      )
   ]
signal time=1681858691.975591 sender=:1.6 -> destination=:1.346 serial=777 path=/org/freedesktop/portal/desktop/request/1_346/t; interface=org.freedesktop.portal.Request; member=Response
   uint32 0
   array [
   ]
error time=1681858691.975602 sender=org.freedesktop.DBus -> destination=:1.6 error_name=org.freedesktop.DBus.Error.ServiceUnknown reply_serial=777
   string "The name :1.346 was not provided by any .service files"
```

</details>

**Comment by unode** [comment]
> So is this why links from Discord open in Chrome instead of my default browser? thinking

I'm having a similar but with `joplin-desktop` and in my case `Chromium` is launched instead despite not even being mentioned in XDG outputs (see below).

I have enabled `xdg.portal` in nixos-22.11 though:

```
  xdg.portal = {
    enable = true;
    xdgOpenUsePortal = true;
    extraPortals = [ pkgs.xdg-desktop-portal-gtk ];
  };
```

`Chromium` is not even among the default http/https handlers:

```
% nix-shell --pure -p glib
$ gio mime x-scheme-handler/https
Default application for ?x-scheme-handler/https?: userapp-Firefox-IN1NE1.desktop
Registered applications:
	userapp-Firefox-IN1NE1.desktop
	google-chrome.desktop
Recommended applications:
	userapp-Firefox-IN1NE1.desktop
	google-chrome.desktop

$ gio mime x-scheme-handler/http                                                                                                                                                                                                                                   
Default application for ?x-scheme-handler/http?: userapp-Firefox-IN1NE1.desktop
Registered applications:
	userapp-Firefox-IN1NE1.desktop
	google-chrome.desktop
Recommended applications:
	userapp-Firefox-IN1NE1.desktop
	google-chrome.desktop
```

Using `gio open http://` opens a new Firefox window, as does `https://`...

Awfully confused about this. :thinking:

**Comment by lilyinstarlight** [comment]
@unode These things run under the `xdg-desktop-portal` systemd user service

Can you see if `systemd-run --user -t gio mime x-scheme-handler/https` correctly shows the default?

If not, you may need to make sure PATH is correctly imported/set for systemd user scope (I did actually consider adding a `systemd.services.xdg-desktop-portal.path = [ "/run/current-system/sw" ]` to the xdg portal module in NixOS, so that the behavior is less surprising by default)

**Comment by unode** [comment]
@lilyinstarlight 
```
% systemd-run --user -t gio mime x-scheme-handler/https
Running as unit: run-u37615.service
Press ^] three times within 1s to disconnect TTY.
Default application for “x-scheme-handler/https”: userapp-Firefox-IN1NE1.desktop
Registered applications:
	userapp-Firefox-IN1NE1.desktop
	firefox.desktop
	chromium-browser.desktop
	google-chrome.desktop
Recommended applications:
	userapp-Firefox-IN1NE1.desktop
	firefox.desktop
	chromium-browser.desktop
	google-chrome.desktop
```
and a similar output for http/https.

It doesn't explain why Firefox isn't launched.
But at least it's clear where Chromium is coming from.

Adding `/run/current-system/sw` to PATH configuration doesn't seem to make a difference.

**Comment by lilyinstarlight** [comment]
@unode If you drop into an interactive shell with `systemd-run --user -t -S`, are you able to get firefox to open with `gio open https://nixos.org/`?

**Comment by unode** [comment]
@lilyinstarlight yes. Both `gio open` and `xdg-open` open a new nixos.org tab in Firefox:

```
% systemd-run --user -t -S
Running as unit: run-u38802.service
Press ^] three times within 1s to disconnect TTY.
% gio open https://nixos.org/
% xdg-open https://nixos.org/
(objectpath '/org/freedesktop/portal/desktop/request/1_38806/t',)
```

**Comment by lilyinstarlight** [comment]
Wait. So xdg-open there (which still goes through portal) opens firefox and xdg-open from a normal shell opens chromium?

**Comment by unode** [comment]
Yes, as far as I can tell `joplin-desktop` uses `xdg-open`. I can see it on an `strace`, however I can't figure out why Firefox isn't launched and Chromium is used instead.

I did however recently have a `x-www-browser` and symlink `www-browser` script in `PATH` that was being executed. Having this script run an arbitrary command worked but having `firefox URL` in it caused Chromium to launch again.

Right now it looks like, for some reason, Firefox fails to launch and Chromium is used as fallback.

**Comment by lilyinstarlight** [comment]
I wonder if joplin-desktop is somehow losing the `NIXOS_XDG_OPEN_USE_PORTAL` environment variable, causing xdg-open to use default logic instead of delegating it to the portal like it's supposed to

**Comment by unode** [comment]
Ok, a few jumps ahead and it turned out that the version of joplin I had on my environment was bundling an unpatched `xdg-open` that ignored `NIXOS_XDG_OPEN_USE_PORTAL`. I thought it was using the system `xdg-open` but I noticed in strace that it was launching `/usr/bin/xdg-open` which after inspection pointed to: `/nix/store/hr3ks9p076js1r8v4p531nml3k0cqsvb-joplin-desktop-2.8.8-usr-target/bin/xdg-open`.

Updated joplin and now Firefox is opening as expected. Confusion mostly cleared though still surprised with the behavior.

@lilyinstarlight Thanks for the help troubleshooting this. Your last comment pointed me in the right direction.

**Comment by huantianad** [comment]
I also was having a similar issue where `xdg-open` wasn't using my defaults. However, the issue ended up unrelated to this. For some reason, bepinex's doorstop was adding the line "Found UnityPlayer, hooking into it instead" to the start of my `.config/mimeapps.list` whenever I launched a game with bepinex modding... I've yet to figure this out but I thought I'd mention it in case it helps someone somehow!

**Comment by LunNova** [comment]
@lilyinstarlight 

> If not, you may need to make sure XDG_DATA_DIRS+PATH is correctly imported/set for systemd user scope

What's the correct way to do this?

I'm seeing this behavior:

```
$ nix shell pkgs#glib -c systemd-run --user -t gio mime x-scheme-handler/http
Running as unit: run-u30.service
Press ^] three times within 1s to disconnect TTY.
No default applications for “x-scheme-handler/http”

$ nix shell pkgs#glib -c gio mime x-scheme-handler/http
Default application for “x-scheme-handler/http”: firefox.desktop
Registered applications:
	firefox.desktop
Recommended applications:
	firefox.desktop
```

**Comment by LunNova** [comment]
Potential workaround for i3 users if your xdg-desktop-portal is failing to start your browser due to path issues:

```
    services.xserver.windowManager.i3.extraSessionCommands = ''
      systemctl --user import-environment PATH
    '';
```

**Comment by lilyinstarlight** [comment]
@LunNova

> What's the correct way to do this?

> ```
> services.xserver.windowManager.i3.extraSessionCommands = ''
>       systemctl --user import-environment PATH
>     '';
> ```

As far as I know, using either `dbus-update-activation-environment` (which updates the systemd user session environment) or `systemctl --user import-environment` is the correct way to do that

I did think about making a PR to add `/run/current-system/sw/bin` to the default `PATH` for the `xdg-desktop-portal` systemd user service to avoid having to import `PATH`, but to be honest importing `PATH` (as well as `XDG_DATA_DIRS`, etc.) in each window manager's config is probably the more "correct" fix and I believe is what GNOME shell does

**Comment by tobiasBora** [comment]
Hum wait, I’m very confused, I just rebooted without adding anything and now it works! Not sure what happened…

Also, any idea why we don’t enable `xdgOpenUsePortal = true;` by default?

**Comment by xNaxdy** [comment]
With `xdgOpenUsePortal = true;`, Lutris is still not properly opening links for me. Running `xdg-open` in `systemd-run --user -t -S` works however.

Any ideas what could cause this ?

**Comment by Slabity** [comment]
Just want to drop in here and note that `xdgOpenUsePortal` completely breaks `xdg-open` on my system. With it enabled, something as simple as `xdg-open https://google.com/` fails to open up firefox, despite `xdg-mime query default x-scheme-handler/https` and `xdg-settings get default-web-browser` both reporting `firefox.desktop`.

I tried to debug it, but honestly there's just no error or debug messages that I could find that led me in the right direction.

As a note, I am using Sway. I'm not sure if there's some startup commands I need to add to get it working properly.

**Comment by Atemu** [comment]
@Slabity which portal do you have configured?

**Comment by Slabity** [comment]
> @Slabity which portal do you have configured?

I'm just gonna post what I think is relevant. Here's what I have for my NixOS configuration:

```nix
programs.sway = {
  enable = true;
  wrapperFeatures.gtk = true;
};

environment.sessionVariables = {
  GTK_USE_PORTAL = "1";
};

xdg.portal = {
  enable = true;
  wlr.enable = true;
  extraPortals = with pkgs; [ xdg-desktop-portal-gtk ];
  #xdgOpenUsePortal = true;
};
```

And for home-manager:

```nix
wayland.windowManager.sway = {
  enable = true;
  systemd.enable = true;
  package = null;
  config = { ... };
};
```

I'm not entirely sure if I need to run some sort of `systemd` or `dbus-*` command to my startup, but I had what was on the unofficial NixOS wiki here for a bit: https://nixos.wiki/wiki/Sway

Though neither the `dbus-sway-environment` or `configure-gtk` parts affected the behavior, so I no longer have them.

My full config can be found here: https://github.com/Slabity/nix-flakes

**Comment by Atemu** [comment]
I don't know how home-manager plays into this. What you need is that the GTK portal user service is running. In DBUS, you should see a service implementing the portal thingies such as file picker.

If portals are not working in general, you can't verify whether this works or not.

**Comment by lilyinstarlight** [comment]
> Though neither the `dbus-sway-environment` or `configure-gtk` parts affected the behavior, so I no longer have them.

You do need something like `dbus-sway-environment` (or at least have sway run `dbus-update-activation-environment --systemd --all`) but make sure it includes all environment variables from your sway config, not just `WAYLAND_DISPLAY` (some like `XDG_DATA_DIRS` and `PATH` are also needed for the portal xdg-open to work correctly)

Ideally we make the NixOS module automatically handle all that, but sway is kinda a diy sort of thing, so portals are not quite functional out-of-the-box in your NixOS config

You can use the command `systemd-run --user -t gio mime x-scheme-handler/https` to test whether it'll work in a portal or not, since it should roughly emulate the environment the portal runs in

**Comment by Slabity** [comment]
Thanks @Atemu and @lilyinstarlight. Calling `dbus-update-activation-environment --systemd --all` does seem to fix everything with `xdgOpenUsePortal` enabled.

I'm not certain, but I *think* there might be an underlying issue in the home-manager sway module. From what I can tell, the NixOS module creates `/etc/sway/config.d/nixos.conf` that contains `exec dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP`, but since my configuration is part of home-manager this config never actually gets executed.

On the other hand, home-manager *does* automatically include the following at the end of the config:

```
exec "/nix/store/29n5vxmqjfy4qips1x66hq6y2m9q0qlp-dbus-1.14.8/bin/dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP XDG_SESSION_TYPE NIXOS_OZONE_WL; systemctl --user start sway-session.target"
```

So it sounds like both the NixOS and home-manager sway modules might not be adding all the necessary environment variables to this list? Presumably the `XDG_DATA_DIRS` and `PATH` ones?

Out of curiosity, is there any issue with just using `--all` instead of listing each variable? Would it be worth updating the NixOS module to use that instead?

**Comment by LunNova** [comment]
> Also, any idea why we don’t enable xdgOpenUsePortal = true; by default?

It could do with some more time to bake, there are too many people running into issues with it.

**Comment by lilyinstarlight** [comment]
> So it sounds like both the NixOS and home-manager sway modules might not be adding all the necessary environment variables to this list? Presumably the `XDG_DATA_DIRS` and `PATH` ones?

Yeah it's missing those necessary variables for xdg-open in portals (portals can otherwise probably work with the variables they have though, e.g. for screenshare or file chooser)

> Out of curiosity, is there any issue with just using `--all` instead of listing each variable? Would it be worth updating the NixOS module to use that instead?

I don't realistically think there are issues, but maybe some people would not prefer that. I've not tried to get that changed. I think that is effectively the same as what DEs like GNOME do though

**Comment by LunNova** [comment]
https://github.com/NixOS/nixpkgs/pull/243834 - this should fix xdgOpenUsePortal with nixos's i3 session.

Once that's in should we consider turning on xdgOpenUsePortal by default and closing this issue?
Are there any other known cases where it doesn't work?

**Comment by huantianad** [comment]
We would need to enable xdg portals in general by default too, right? Is there any issue with that?

**Comment by lilyinstarlight** [comment]
Perhaps particular desktop environments can just enable it themselves when the module is compatible with it?

For example the GNOME and Plasma NixOS modules and I guess now i3 could enable it by default (once your PR is merged), but for anyone who is using a module that will not support it out-of-the-box (e.g. sway) or is putting together their own DE/WM without a NixOS module and does not having portals fully-working, enabling it by default would just lead to very surprising behavior with no useful warnings or error messages to indicate what is wrong

> We would need to enable xdg portals in general by default too, right? Is there any issue with that?

The `xdgOpenUsePortal` option only has any effect when portals themselves are also enabled (but as mentioned, just because they are enabled don't mean they are fully working or that the needed portals are added)

**Comment by unode** [comment]
Sorry to highjack the thread but I seem to be encountering issues with Joplin Desktop again. Opening links in Firefox works (as discussed above) but opening an embedded PDF in an external app doesn't.

The PATH update suggestions mentioned in another ticket didn't have an effect.

It's also difficult to debug as I can't seem to replicate the behavior with `systemd-run --user -t`. In that case, when opening a PDF, I get a dialog asking which program to use (Evince, Zathura, ...)

How are you troubleshooting these issues and making sure that the cause of the problem is not a different one?

**Comment by dyfrgi** [comment]
Freedesktop deleted their xdg-utils repo here so I can't see whether https://github.com/freedesktop/xdg-utils/pull/12 was declined or just not merged. I filed https://gitlab.freedesktop.org/xdg/xdg-utils/-/issues/240 asking upstream to implement a similar override. Thanks for this work. Perhaps this should be turned on by default for home-manager when running in an FHS env as I expect this will be a problem for many apps installed in that way.

**Comment by newzubakhin** [comment]
@larsivi 
> I then took the url that was opened in Firefox, copied it over in Vivaldi, then approved it there. This then opened up an xdg-open dialog (the link is jetbrains://) which when clicked on does absolutely nothing.

As I can see, [Jetbrains Toolbox from nixpks](https://github.com/NixOS/nixpkgs/blob/0b3d618173114c64ab666f557504d6982665d328/pkgs/applications/misc/jetbrains-toolbox/default.nix#L35) doesn't handle `jetbrains://` links at all. You can check `.desktop` in the archive from the official site and copy `MimeType=x-scheme-handler` section from it.

**Comment by Atemu** [comment]
That looks like a fairly trivial fix on the Nixpkgs side.

**Comment by Prn-Ice** [comment]
For anyone with the KDE/vscodeFHS issue. I think you can fix it with some of the suggestions here. Remember to restart your PC after updating the portal settings.

What worked for me was
- Add this to your system nix configs
  ```nix
  xdg.portal = {
      enable = true;
      xdgOpenUsePortal = true;
    };
  ```

- Install `xdg-utils` with home-manager

**Comment by LunNova** [comment]
@K900 thoughts on changing the plasma 6 defaults like @viscountexx suggested? We could set the env var only in the plasma 6 session so it doesn't impact non-plasma sessions.

**Comment by K900** [comment]
We could mkDefault it I guess, but I'd be very careful with the potential side effects.

**Comment by sbuller** [comment]
So, I just stumbled on this while debugging an issue with a URL handler. The xdgOpenUsePortal definitely deals with the issue, but I end up with a much bigger problem. All of a sudden xdg-open seems to have a split personality. I call `xdg-mime query default inode/directory`, and it (properly) tells me `org.kde.dolphin.desktop`, but when I `xdg-open ~` I get VSCodium trying to open the directory. I have since discovered that `gio mime inode/directory` (mentioned above) simply lists the desktop files it finds in alphabetical order, rather than consulting the preferences set in KDE. My reading of this is that xdg-open, part way through, is forgetting that it's running on KDE.

**Comment by sbuller** [comment]
So, I decided to go the route of making .desktop files executable, and made an overlay to handle the bulk of it. It doesn't work with every package, but it seems to work for many.

```nix
  executableDesktopFiles = final: _prev: let
    pkgs = import inputs.nixpkgs {
      system = "x86_64-linux";
      config.allowUnfree = true;
    };
    fixDesktop = name: pkg: pkgs.runCommand "${name}-exedesk" {} ''
      mkdir -p $out/share/applications
      cp -ur ${pkg.out}/* $out/
      cp -nL ${pkg.out}/share/applications/* $out/share/applications || true
      for f in $out/share/applications/*; do
        target=$(readlink -f "$f")
        if [ -L "$f" ]; then
          rm "$f"
          cp "$target" "$f"
        fi
        chmod +x "$f"
      done
    '';
  in builtins.mapAttrs fixDesktop {
    inherit (pkgs)
      audacity
      blender
      brave
      ...
    ;
  };
```

**Comment by deftdawg** [comment]
Based on the discussion here and the fact Firefox (with an executable .desktop file which brave doesn't have) works perfectly, while Joplin-Desktop opening links -> Brave browser does this:
![image](https://github.com/NixOS/nixpkgs/assets/4991612/bf7e25c9-9f7a-4910-b9f8-2db001d54cdc)
![image](https://github.com/NixOS/nixpkgs/assets/4991612/3d2408e2-30b3-4c17-86ba-6446c9ec0332)

I'm having this same challenge...  I'll give @sbuller's overlay a try and hope for a less hacky fix to this issue to eventually emerge.  :beers:

**Comment by deftdawg** [comment]
Tried to incorporate @sbuller's overlay into my `/etc/nixos/configuration.nix`, unfortunately it didn't work as is...  My brain is a bit cooked (long day), I tweaked the config into this which produces an infinite recursion error.  :sleepy: 

```nix
 # MARK: Joplin-Desktop -> Brave URL Open fix
 nixpkgs.overlays = [
    (final: _prev: let
      # pkgs = import inputs.nixpkgs {
      #   system = "x86_64-linux";
      #   config.allowUnfree = true;
      # };
      fixDesktop = name: pkg: pkgs.runCommand "${name}-exedesk" {} ''
        mkdir -p $out/share/applications
        cp -ur ${pkg.out}/* $out/
        cp -nL ${pkg.out}/share/applications/* $out/share/applications || true
        for f in $out/share/applications/*; do
          target=$(readlink -f "$f")
          if [ -L "$f" ]; then
            rm "$f"
            cp "$target" "$f"
          fi
          chmod +x "$f"
        done
      '';
    in builtins.mapAttrs fixDesktop {
      inherit (pkgs)
      brave;
    }
    )
  ];
```

**Comment by sbuller** [comment]
Sorry about that whole `let pkgs=...` business, I was trying to deal with infinite recursion myself and it took a bit of doing before I got something working. You should replace `pkgs` and `_prev` with `prev`. To clarify, the underscore on `_prev` is just there to show it's not being used currently and should be removed when you start using `prev` in place of `pkgs`.

**Comment by LunNova** [comment]
If that fixes launching you're probably launching the handler for that filetype inside the FHS bubblewrap container for the app you tried to open it from, so it's not a great solution.

https://github.com/NixOS/nixpkgs/issues/107826 decided making them executable was a bad idea.

**Comment by deftdawg** [comment]
> If that fixes launching you're probably launching the handler for that filetype inside the FHS bubblewrap container for the app you tried to open it from, so it's not a great solution.
> 
> #107826 decided making them executable was a bad idea.

Yep, it's bad, I don't disagree... however that issue is also from 4 years ago and this issue is over 2 years old.   This problem is annoying, I'll live with the risks using the workaround for it until there's a proper fix.

Here's the full working code for @sbuller's workaround for anyone else that wants:
```nix
  # MARK: Joplin-Desktop -> Brave URL open workaround (credit: sbuller)
  nixpkgs.overlays = [
    (final: prev: let
      fixDesktop = name: pkg: prev.runCommand "${name}-exedesk" {} ''
        mkdir -p $out/share/applications
        cp -ur ${pkg.out}/* $out/
        cp -nL ${pkg.out}/share/applications/* $out/share/applications || true
        for f in $out/share/applications/*; do
          target=$(readlink -f "$f")
          if [ -L "$f" ]; then
            rm "$f"
            cp "$target" "$f"
          fi
          chmod +x "$f"
        done
      '';
    in builtins.mapAttrs fixDesktop {
      inherit (prev)
      brave;
    }
    )
  ];
```

**Comment by sbuller** [comment]
I'm really struggling with debugging DBus here. Is it possible that kde / xdg-desktop-portal-kde just doesn't implement OpenURI, and when gdbus tries to call it gdbus just falls back to gio which then uses a default gnome config?

**Comment by LunNova** [comment]
The KDE portal should support OpenURI.

What do you have set under xdg.portal? There's a `config` attrset under it which controls which portals are used under which desktop environments. Maybe that priority is set up incorrectly :thinking: 

Here's an example which sets it for KDE and i3 in case that helps.

```nix
    xdg.portal = {
      enable = true;
      xdgOpenUsePortal = true;
      # I think the name under config has to match up with the value of XDG_CURRENT_DESKTOP.
      config.KDE.default = [ "kde" "*" ]; # kde portal by default, fallback to anything
      config.${"none+i3"}.default = [ "kde" "*" ];
      extraPortals = [
        pkgs.xdg-desktop-portal-gtk
        pkgs.xdg-desktop-portal-wlr
      ];
    };
```

**Comment by sbuller** [comment]
I'd had
```nix
xdg.portal = {
  enable = true;
  xdgOpenUsePortal = true;
  wlr.enable = true;
  extraPortals = [ pkgs.xdg-desktop-portal-kde ];
};
```
but no variation that I've tried has made any difference, including an exact copy of the above. In every instance `xdg-open ~` uses vscodium, while kde-open uses dolphin.

I went and looked through https://github.com/KDE/xdg-desktop-portal-kde, and found no mention of OpenURI, but I'm not sure if that's the right place to look.

**Comment by sbuller** [comment]
Explicitly setting xdg.mime.{addedAssociations,defaultApplications} works around the issue for the mimetypes which I've set. This is unsatisfactory however, as I don't get to benefit from the long list of associations set by my desktop environment.

**Comment by lilyinstarlight** [comment]
Judt a heads-up, we should probably update xdgUsePortal to use the env vars from <https://gitlab.freedesktop.org/xdg/xdg-utils/-/merge_requests/112> if it is merged (seems likely) and remove our own patch

**Comment by ilya-fedin** [comment]
> I'm really struggling with debugging DBus here. Is it possible that kde / xdg-desktop-portal-kde just doesn't implement OpenURI, and when gdbus tries to call it gdbus just falls back to gio which then uses a default gnome config?

Just to clear it, the portal doesn't ask backends for the default application, it always uses glib to retrieve the default application and the list of applications. Backends are called only when the frontend wants to present app chooser dialog and backends get the apps list from the frontend which in turn gets it from glib.

In other words, when you let portal handle your application defaults, you in fact let glib handle them. Which is not a bad thing I would say as xdg-open's offload to DE binaries with custom MIME handling is beating desktop Linux for a long time. While glib always follows mime-apps-spec.

Plasma 5 used a custom way to handle initial defaults (`InitialPreference` key in .desktop files) while Plasms 6 follows mime-apps-spec in this regard. With Plasma 6, the existing of `/run/current-system/sw/share/applications/kde-mimeapps.list` and xdg-desktop-portal having `XDG_CURRENT_DESKTOP=KDE` in its environment should be enough for proper defaults to be used (could be checked with `sudo cat /proc/$(pgrep -f /xdg-desktop-portal\$)/environ | tr '\0' '\n' | grep XDG_CURRENT_DESKTOP`).

**Comment by eclairevoyant** [comment]
> Just a heads-up, we should probably update xdgUsePortal to use the env vars from https://gitlab.freedesktop.org/xdg/xdg-utils/-/merge_requests/112 if it is merged (seems likely) and remove our own patch

This MR is merged as of a year ago. We should now use those envvars.

**Comment by sliedes** [comment]
I think one part of the puzzle is that some browsers (like google-chrome) have absolute paths in their `Exec=` lines while others don't. Particularly, Chrome uses an absolute path while Firefox does not. This, I believe, means that Chrome will work even when `PATH` is not correctly set for the portal. Apparently the paths should **not** be absolute (see https://github.com/NixOS/nixpkgs/issues/308324).

Furthermore, I think sway doesn't update `PATH` (not filing an issue because it might may do it elsewhere; I'm actually using home-manager config myself, which _does_ seem to have the same issue). I do not know if this is an oversight or if there is some disagreement about whether `PATH` should be updated.

https://github.com/NixOS/nixpkgs/blob/4ba651b7442cba2ef940b6a1d4386a8396d20c9a/nixos/modules/programs/wayland/sway.nix#L162

As a practical result, clicking links in gnome-terminal on my sway session opens them in Chrome while almost everything else opens them in Firefox (no flatpak or FHS involved).

**Comment by Ralith** [comment]
Discussed this at some length in Matrix tonight, and we concluded that passing `PATH` into the systemd user unit environment is probably the right solution. https://github.com/nix-community/home-manager/pull/8524 will fix this for sway in home-manager; other environments will likely need a similar fix.

**Comment by eclairevoyant** [comment]
That's a terrible idea, a user's PATH is full of all sorts of binaries that should not be part of a systemd service without explicit opt-in.

**Comment by eclairevoyant** [comment]
Systemd units are designed to be isolated, as evidenced by the fact that they don't inherit `PATH` in the first place. Having the sway module shatter that expectation would be extremely surprising and overall questionable - the flatpak people's apparent ignorance of systemd notwithstanding.

**Comment by sliedes** [comment]
> Systemd units are designed to be isolated, as evidenced by the fact that they don't inherit `PATH` in the first place. Having the sway module shatter that expectation would be extremely surprising and overall questionable - the flatpak people's apparent ignorance of systemd notwithstanding.

I think you are confusing mechanism and intended effect here. systemd is a way to run code in a controlled environment, yes, puncturing it exactly where you need to.

For the xdg portal, my understanding is that it using PATH is a necessary feature, not a bug. It is *supposed* to be a controlled way to, among other things, run your .desktop files *in your current environment*. In essence, instead of a software being able to say "execute firefox" or "execute xdg-open" (which would inherently, and properly, use PATH—in fact, the whole point of PATH is that the user can change what is executed, isn't it?), it can signal the portal "open the user-approved browser in the user's environment".

You seem to place a lot of meaning to the fact that it's run as a systemd service. But 1) it doesn't need to be; that's just a convenient way to do it; 2) when it is, it still needs to be provided the environment it needs to accomplish its very purpose, executing software in the user's environment (DISPLAY, PATH, ...).

So, yes, you are right: systemd provides isolation for a reason. It's a sensible default. And even here I think it would be bad to blindly pass everything to it.

But, xdg-portal fundamentally **is a tool to execute things in your PATH**, by design. The only big difference is that it can also be used by software that is not allowed to directly execute commands in your PATH.

See, it is completely legitimate for the user to write a .desktop file that says "I want xterm executed", with the desired semantics of executing xterm that is in his PATH. Given that file, it would be broken behavior to 1) execute xterm from somewhere else, or 2) fail to execute xterm.

And the way xdg-portal, by design, accomplishes this is that it searches the user's PATH for the correct binary to execute. The fact that it's a systemd user service is a complete red herring.

**Comment by sliedes** [comment]
> I never implied the portal had to be a service. I'm concerned about the suggestion to pollute the entire systemd user session over one service, which itself is an impl detail.

Wait, does this actually make the PATH part of *all* systemd user sessions? If that is the case, I can see your point.
